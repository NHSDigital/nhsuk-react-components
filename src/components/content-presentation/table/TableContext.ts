import { createContext } from 'react';

export interface ITableContext {
  isResponsive: boolean;
  headings: string[];
  setHeadings(headings: string[]): void;
}

const TableContext = createContext<ITableContext>({
  /* eslint-disable @typescript-eslint/no-empty-function */
  isResponsive: false,
  headings: [],
  setHeadings: () => {},
});

export default TableContext;
