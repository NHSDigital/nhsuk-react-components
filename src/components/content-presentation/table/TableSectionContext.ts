import { createContext } from 'react';

export enum TableSection {
  NONE,
  HEAD,
  BODY,
}

// consider instead of enum
// const TableSection = {
//   NONE: 'NONE',
//   HEAD: 'HEAD',
//   BODY: 'BODY'
// } as const

const TableSectionContext = createContext<TableSection>(TableSection.NONE);

export default TableSectionContext;
