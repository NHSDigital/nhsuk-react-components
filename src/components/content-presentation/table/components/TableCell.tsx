import classNames from 'classnames';
import React, { FC, HTMLProps, useContext } from 'react';
import useDevWarning from '@util/hooks/UseDevWarning';
import TableSectionContext, { TableSection } from '../TableSectionContext';

const CellOutsideOfSectionWarning =
  'Table.Cell used outside of a Table.Head or Table.Body component. Unable to determine section type from context.';

export interface TableCellProps extends HTMLProps<HTMLTableCellElement> {
  _responsive?: boolean;
  _responsiveHeading?: string;
  isNumeric?: boolean;
}

const TableCell: FC<TableCellProps> = ({
  className,
  _responsive = false,
  _responsiveHeading = '',
  isNumeric,
  children,
  ...rest
}) => {
  const section = useContext(TableSectionContext);
  useDevWarning(CellOutsideOfSectionWarning, () => section === TableSection.NONE);

  const cellClass = section === TableSection.HEAD ? 'nhsuk-table__header' : 'nhsuk-table__cell';
  const classes = classNames(cellClass, { [`${cellClass}--numeric`]: isNumeric }, className);

  return (
    <>
      {section === TableSection.HEAD ? (
        <th className={classes} scope="col" {...rest}>
          {children}
        </th>
      ) : (
        <td className={classes} role={_responsive ? 'cell' : undefined} {...rest}>
          {_responsive && (
            <span className="nhsuk-table-responsive__heading">{_responsiveHeading}</span>
          )}
          {children}
        </td>
      )}
    </>
  );
};

TableCell.displayName = 'Table.Cell';

export default TableCell;
