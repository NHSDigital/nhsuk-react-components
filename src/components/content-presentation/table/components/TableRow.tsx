'use client';
import classNames from 'classnames';
import React, { Children, cloneElement, FC, HTMLProps, useContext, useEffect } from 'react';
import TableContext from '../TableContext';
import { getHeadingsFromChildren, isTableCell } from '../TableHelpers';
import TableSectionContext, { TableSection } from '../TableSectionContext';

const TableRow: FC<HTMLProps<HTMLTableRowElement>> = ({ className, children, ...rest }) => {
  const section = useContext(TableSectionContext);
  const { isResponsive, headings, setHeadings } = useContext(TableContext);

  useEffect(() => {
    if (isResponsive && section === TableSection.HEAD) {
      setHeadings(getHeadingsFromChildren(children));
    }
  }, [isResponsive, section, children]);

  if (isResponsive && section === TableSection.BODY) {
    const tableCells = Children.map(children, (child, index) => {
      if (isTableCell(child)) {
        return cloneElement(child, {
          _responsive: isResponsive,
          _responsiveHeading: `${headings[index] || ''} `,
        });
      }
      return child;
    });

    return (
      <tr className={classNames('nhsuk-table__row', className)} {...rest}>
        {tableCells}
      </tr>
    );
  }

  return (
    <tr className={classNames('nhsuk-table__row', className)} {...rest}>
      {children}
    </tr>
  );
};
TableRow.displayName = 'Table.Row';

export default TableRow;
