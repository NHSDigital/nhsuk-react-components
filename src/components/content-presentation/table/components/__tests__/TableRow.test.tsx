import { render } from '@testing-library/react';
import React from 'react';
import TableContext, { ITableContext } from '../../TableContext';
import TableSectionContext, { TableSection } from '../../TableSectionContext';
import TableCell from '../TableCell';
import TableRow from '../TableRow';

const assertCellText = (container: HTMLElement, cellNumber: number, text: string) => {
  expect(container.querySelector(`[data-test="cell-${cellNumber}"]`)?.textContent).toEqual(text);
};

describe('Table.Row', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <table>
        <tbody>
          <TableRow />
        </tbody>
      </table>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders headers in the first column if responsive', () => {
    const contextValue: ITableContext = {
      isResponsive: true,
      headings: ['a', 'b', 'c'],
      setHeadings: jest.fn(),
    };
    const { container } = render(
      <table>
        <TableContext.Provider value={contextValue}>
          <TableSectionContext.Provider value={TableSection.BODY}>
            <tbody>
              <TableRow>
                <TableCell data-test="cell-1">1</TableCell>
                <TableCell data-test="cell-2">2</TableCell>
                <TableCell data-test="cell-3">3</TableCell>
              </TableRow>
            </tbody>
          </TableSectionContext.Provider>
        </TableContext.Provider>
      </table>,
    );

    assertCellText(container, 1, 'a 1');
    assertCellText(container, 2, 'b 2');
    assertCellText(container, 3, 'c 3');
    expect(container.querySelectorAll('.nhsuk-table-responsive__heading').length).toBe(3);
  });

  it('renders row contents without headers in responsive mode if they are not cells', () => {
    const contextValue: ITableContext = {
      isResponsive: true,
      headings: ['a', 'b', 'c'],
      setHeadings: jest.fn(),
    };
    const { container } = render(
      <table>
        <TableContext.Provider value={contextValue}>
          <TableSectionContext.Provider value={TableSection.BODY}>
            <tbody>
              <TableRow>
                <td data-test="cell-1">1</td>
                <td data-test="cell-2">2</td>
                <td data-test="cell-3">3</td>
              </TableRow>
            </tbody>
          </TableSectionContext.Provider>
        </TableContext.Provider>
      </table>,
    );

    assertCellText(container, 1, '1');
    assertCellText(container, 2, '2');
    assertCellText(container, 3, '3');
    expect(container.querySelectorAll('.nhsuk-table-responsive__heading').length).toBe(0);
  });

  it('renders row contents as headers in head section in responsive mode', () => {
    const setHeadings = jest.fn();
    const contextValue: ITableContext = {
      isResponsive: true,
      headings: ['a', 'b', 'c'],
      setHeadings,
    };
    render(
      <table>
        <TableContext.Provider value={contextValue}>
          <TableSectionContext.Provider value={TableSection.HEAD}>
            <thead>
              <TableRow>
                <TableCell data-test="cell-1">1</TableCell>
                <TableCell data-test="cell-2">2</TableCell>
                <TableCell data-test="cell-3">3</TableCell>
              </TableRow>
            </thead>
          </TableSectionContext.Provider>
        </TableContext.Provider>
      </table>,
    );

    expect(setHeadings).toHaveBeenCalledWith(['1', '2', '3']);
  });

  it('sets headers, skipping contents outside of table cells in responsive mode', () => {
    const setHeadings = jest.fn();
    const contextValue: ITableContext = {
      isResponsive: true,
      headings: ['a', 'b', 'c'],
      setHeadings,
    };
    render(
      <table>
        <TableContext.Provider value={contextValue}>
          <TableSectionContext.Provider value={TableSection.HEAD}>
            <thead>
              <TableRow>
                <td data-test="cell-1">1</td>
                <TableCell data-test="cell-2">2</TableCell>
                <td data-test="cell-3">3</td>
              </TableRow>
            </thead>
          </TableSectionContext.Provider>
        </TableContext.Provider>
      </table>,
    );

    expect(setHeadings).toHaveBeenCalledWith(['2']);
  });

  it('does not render row contents as headers in head section in normal mode', () => {
    const contextValue: ITableContext = {
      isResponsive: false,
      headings: ['a', 'b', 'c'],
      setHeadings: jest.fn(),
    };
    const { container } = render(
      <table>
        <TableContext.Provider value={contextValue}>
          <TableSectionContext.Provider value={TableSection.HEAD}>
            <thead>
              <TableRow>
                <TableCell data-test="cell-1">1</TableCell>
                <TableCell data-test="cell-2">2</TableCell>
                <TableCell data-test="cell-3">3</TableCell>
              </TableRow>
            </thead>
          </TableSectionContext.Provider>
        </TableContext.Provider>
      </table>,
    );

    assertCellText(container, 1, '1');
    assertCellText(container, 2, '2');
    assertCellText(container, 3, '3');
    expect(container.querySelectorAll('.nhsuk-table-responsive__heading').length).toBe(0);
  });
});
