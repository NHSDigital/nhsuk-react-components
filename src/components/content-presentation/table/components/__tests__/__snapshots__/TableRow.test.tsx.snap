// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.Row matches snapshot 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="nhsuk-table__row"
      />
    </tbody>
  </table>
</div>
`;
